doc/design/logging.md

```diff
@@ -0,0 +1,396 @@
+# Logging
+
+## Summary and problem description
+
+Logging is a crucial part of a framework since it informs the developer and
+user of the framework about anomalies and helps with debugging by providing
+contextual information.
+
+Additionally, when integrated into a separate framework, the log messages should
+be forwarded in order to have a single logging infrastructure.
+
+The logging should also be performant and not allocate memory. Ideally it should
+be possible to disable it at compile time and let the compiler optimize it away.
+
+It should also be possible to alter the behavior of the logging via environment
+variables, e.g. the log level.
+
+## Terminology
+
+### Log levels and how to use them
+
+| LogLevel | Usage |
+|----------|-------|
+| FATAL    | For fatal and non-recoverable errors which essentially lead to termination of the program. |
+| ERROR    | For severe but recoverable errors. |
+| WARN     | For cases when something unintentional happens which is not considered a severe error. |
+| INFO     | For important status information a user should be aware of. |
+| DEBUG    | Anything that is helpful for debugging. |
+| TRACE    | Any comment could be replaced by logging call with log level trace. |
+
+The log levels `FATAL`, `ERROR` and `WARN` should not be used independently but
+in combination with the error handler.
+
+## Design
+
+### Considerations
+
+1. Similar to other iceoryx constructs, heap usage, exceptions and recursive
+   calls are forbidden. This also means that `std::string` and
+   `std::stringstream` cannot be used.
+
+2. Since the logger shall be used by the `iceoryx_hoofs` itself, the base
+   implementation can also not make use of functionality in e.g. `iox::cxx` or
+   `iox::posix`
+
+3. To have a minimal runtime overhead, string formatting shall only happen when
+   the log level is above the output threshold. This lazy evaluation shall also
+   apply to other potentially expensive operations like the execution of a
+   function call.
+
+4. Additionally, the minimal log level shall be configurable by a compile time
+   switch and everything below this log level should be compiled to no-ops. With
+   this, developer can make excessive use of the most verbose log level and
+   disable it by default. When debugging, this can be turned on and help to find
```

Comment:
@elfenpiff this is two fold. With the compile time switch, the tests will ensure that the behavior is the same, therefore I don't see a big issue with this. The log level can also be set at runtime and yes this would influence the runtime behavior but I think it would be a bad idea to always pay the price for a log message even if it will be discarded. It the user wants the same behavior he just needs to use the same compiler time switch and don't allow to change the log level (e.g. via environment variables), which is possible with the current proposal.